### PR TITLE
Starting 3d rot shift convenience layer

### PR DESCRIPTION
**Pull request by sjrothfuss**
Hey! Working on #10 style convenience layer for 3D images. I've got a function working but I want to call something to your attention before merging. To get an x rotation, I have to pass it to `torch_affine_utils.transforms_3d.Rz`. Can anyone help me understand why?

Here's a test image with a dot:
image

When I rotate that with a value in the third position, it rotates about x:
image

As shown in these lines, the value in the third position is sent to `Rz`. https://github.com/teamtomo/torch-transform-image/pull/12/files#diff-99cabfec4ebe678fd3f7e928679a277924006701dece61ad561e6c561e81912aR73-R80

TL;DR it works but I don't know why. :D

### CODE REVIEW

**Comment by jojoelfe** [comment]
Are we looking at a XY-plane of a 3D volume? If yes, than my interpretation is that this volume was rotated around the Z-axis, which should happen with Rz.

**Comment by sjrothfuss** [comment]
@jojoelfe Ready for review! Both `shift_then_rotate` and `rotate_then_shift` are working with tests.

One note on referring to shift axes. I followed the [docstring in `torch_affine_utils.transforms_3d.T`](https://github.com/teamtomo/torch-affine-utils/blob/main/src/torch_affine_utils/transforms_3d.py#L204-L214) that says the order is (x, y, z) even though (n, 0, 0) shows as a vertical change in matplotlib `imshow`. Feedback on that is especially welcomed.

File: src/torch_transform_image/transforms_3d.py [hunk 1]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+        Rz(rotate_tensor[2]) @
+        T(shift_tensor) @
+        T(-center_tensor)
+    )
+
+    return affine_transform_image_3d(
+        image=image,
+        matrices=matrix,
+        interpolation=interpolation,
+        zyx_matrices=True,
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:120]:
```suggestion
        zyx_matrices=False,
```

These should be set to False since you have written the matrices to work on xyz coordinates :+1:

File: src/torch_transform_image/transforms_3d.py [hunk 2]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+        Ry(rotate_tensor[1]) @
+        Rz(rotate_tensor[2]) @
+        T(-center_tensor)
+    )
+
+    return affine_transform_image_3d(
+        image=image,
+        matrices=matrix,
+        interpolation=interpolation,
+        zyx_matrices=True,
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:204]:
```suggestion
        zyx_matrices=False,
```

Same here

**Review by McHaillet — CHANGES_REQUESTED** [review]
Actually the tests results also got me very confused! I have been looking at it for quite some time.

Small example to show what I would expect from your test:

```python
c = torch.tensor([[14, 14, 8, 1]]).float()
c = einops.rearrange(c, 'b xyzw -> b xyzw 1')
torch.linalg.inv(T([[14, ] * 3]) @ Rz(90) @ T([[-5, 0, 0]]) @ T([[-14, ] * 3])) @ c
>>>
tensor([[[19.0000],
         [14.0000],
         [ 8.0000],
         [ 1.0000]]])
```
So I'd expect to see your point at (19, 14, 8) (xyz), or in image coordinates at result[8, 14, 19]. 

The linalg.inv is perhaps a bit confusing here! But I do it because when multiplying the grid point with the transform, we calculate **where it will interpolate data from**. By inverting I calculate where the point actually moves to!

_edit becaue I made a mistake in copying_

File: src/torch_transform_image/transforms_3d.py [hunk 3]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+        raise ValueError(
+            f"3 angles are required but {num_angles} were supplied: {rotate_xyz}."
+        )
+    if (num_shifts := shift_tensor.numel()) != 3:
+        raise ValueError(
+            f"3 shifts are required but {num_shifts} were supplied: {shifts_xyz}."
+        )
+
+    matrix = (
+        T(center_tensor) @
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:108]:
```suggestion
        T(center_tensor) @
        T(shift_tensor) @
```

File: src/torch_transform_image/transforms_3d.py [hunk 4]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+        Rx(rotate_tensor[0]) @
+        Ry(rotate_tensor[1]) @
+        Rz(rotate_tensor[2]) @
+        T(shift_tensor) @
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:112]:
```suggestion
```

File: src/torch_transform_image/transforms_3d.py [hunk 5]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+            f"3 angles are required but {num_angles} were supplied: {rotate_xyz}."
+        )
+    if (num_shifts := shift_tensor.numel()) != 3:
+        raise ValueError(
+            f"3 shifts are required but {num_shifts} were supplied: {shifts_xyz}."
+        )
+
+    matrix = (
+        T(center_tensor) @
+        T(shift_tensor) @
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:193]:
```suggestion
```

File: src/torch_transform_image/transforms_3d.py [hunk 6]
```diff
@@ -38,3 +37,169 @@ def affine_transform_image_3d(
...
+        Rx(rotate_tensor[0]) @
+        Ry(rotate_tensor[1]) @
+        Rz(rotate_tensor[2]) @
```
Comment (McHaillet) [inline src/torch_transform_image/transforms_3d.py:196]:
```suggestion
        Rz(rotate_tensor[2]) @
        T(shift_tensor) @
```

**Review by McHaillet — COMMENTED** [review]
I think you also need to inverse the order of the rotation and translation to correspond with your function names. The matrices are premultiplied, so the right most matrix does the first operation

**Comment by McHaillet** [comment]
Sorry for all the comments, really nice work that you are implementing these functions!

**Comment by alisterburt** [comment]
@sjrothfuss great work here - I saw @McHaillet commented that there was some xyz ordering: let's stick to zyx everywhere internally here

**Comment by McHaillet** [comment]
> I think you also need to inverse the order of the rotation and translation to correspond with your function names. The matrices are premultiplied, so the right most matrix does the first operation. Not 100% sure about that, matrix multiplications are always confusing to think about.
> 
> _edit_
> 
> ```python
> c = torch.tensor([[14, 14, 8, 1]]).float()
> c = einops.rearrange(c, 'b xyzw -> b xyzw 1')
> T([[14, ] * 3]) @ Rz(90) @ T([[-5, 0, 0]]) @ T([[-14, ] * 3]) @ c
> >>>  # this adds -5 to x and then rotates 90 around z, hence the 9 ends up in y
> tensor([[[14.0000],
>          [ 9.0000],
>          [ 8.0000],
>          [ 1.0000]]])
> 
> T([[14, ] * 3]) @ T([[-5, 0, 0]]) @ Rz(90) @ T([[-14, ] * 3]) @ c
> >>>  # this first rotates 90 around z (nothing happens for 14,14), and then adds -5 to x
> tensor([[[ 9.0000],
>          [14.0000],
>          [ 8.0000],
>          [ 1.0000]]])
> ```

This might actually be BS, because here I should have also applied the torch.linalg.inv to calculate where the points ends up.
And then it does behave as rotation followed by translation:
```python
c = torch.tensor([[14, 14, 8, 1]]).float()
c = einops.rearrange(c, 'b xyzw -> b xyzw 1')
torch.linalg.inv(T([[14, ] * 3]) @ Rz(90) @ T([[-5, 0, 0]]) @ T([[-14, ] * 3])) @ c
>>>
tensor([[[19.0000],
         [14.0000],
         [ 8.0000],
         [ 1.0000]]])
```

**Comment by jojoelfe** [comment]
Here is a test notebook, I made from your PR:

https://nbviewer.org/urls/gist.githubusercontent.com/jojoelfe/3fc8c32e7064a4b015f2d753f2138c48/raw/35fec6249274b0c5bbf2b5198a71a25a3a12ab8b/test.ipynb

In this example I am rotating the cone by 30 deg around Y (blue is unrotated, green is rotated). To me it looks like the rotation is following the left-hand rule, whereas we should follow the right-hand rule.

**Comment by sjrothfuss** [comment]
I've uploaded a notebook with examples, see it at [examples/rotate_shift_image_3d_examples.ipynb](https://github.com/sjrothfuss/torch-transform-image/blob/3d-rot-shift-convenience-layer/examples/rotate_shift_image_3d_examples.ipynb). (I tried an [nbviewer rendering](https://nbviewer.org/github/sjrothfuss/torch-transform-image/blob/3d-rot-shift-convenience-layer/examples/rotate_shift_image_3d_examples.ipynb) but it's still showing an out of date version. ☹️)

## Issue 1: Order of operations

@McHaillet I think you caught your error here, I'm pretty confident I have the correct order, see the examples notebook.

## Issue 2: Axis order

I think I found the issue. I had been assuming that when I set `image[8, 14, 14] = 1`, I was modifying the point at `x=8, y=14, z=14` so that the point was off center in x. Thanks to @jojoelfe for pointing out that actually sets `z=8, y=14, x=14`. You can look at the example notebook and see that axis order now appears to be zyx for shifts and rotations. 🎉 It also looks to follow the right hand rule, let me know if you see something different @jojoelfe.

However, I'm still passing the first element of `rotate_zyx` to `Rx`! See [these lines](https://github.com/sjrothfuss/torch-transform-image/blob/3d-rot-shift-convenience-layer/src/torch_transform_image/transforms_3d.py#L109-L111). Can anyone explain why I'm not seeing x rotation then?

**Comment by McHaillet** [comment]
> However, I'm still passing the first element of `rotate_zyx` to `Rx`! See [these lines](https://github.com/sjrothfuss/torch-transform-image/blob/3d-rot-shift-convenience-layer/src/torch_transform_image/transforms_3d.py#L109-L111). Can anyone explain why I'm not seeing x rotation then?

As I mentioned before, this is because you are setting the `zyx_matrices` keyword argument to True in your call to affine_transform_image_3d! If you apply the Rx matrix from torch-affine-utils, which by default works on xyz-coordinates, to zyx ordered coordinates, it is basically like applying an Rz matrix. 

So if you should set `zyx_matrices` to False, then you will get the desired result (at least for the rotations)

**Comment by sjrothfuss** [comment]
FYI @jojoelfe specifying `zyx=True` for the rotation matrices also reverses the rotation direction.
image

**Comment by alisterburt** [comment]
xyz is the sane default to me, allows for the simple case (transforming xyz coordinates) without the overhead of understanding numpy/C style memory layout/axis ordering conventions.

Progressive disclosure of complexity, only introduce it where necessary :-)

**Comment by McHaillet** [comment]
> FYI @jojoelfe specifying `zyx=True` for the rotation matrices also reverses the rotation direction. image

sorry, accidentally edited your post instead of making a new comment. I removed my edit, should be your original message again. I wanted to say:

That's what I mean with composing the forward transform matrix, and then using linalg.inv when you pass it to affine transform. Then you shouldn't have that problem.

**Comment by alisterburt** [comment]
yep I agree - I would probs argue for axis angle or a rotation matrix
